File: python/double_linked_list/double_linked_list.py
import random

class Node:
  def __init__(self, v, nxt, prev):
    self.v = v
    self.nxt = nxt
    self.prev = prev

class DoubleLinkedList:
  def __init__(self):
    self.head = None

  def addSorted(self, v):
    temp = Node(v, None, None)
    if not self.head:
      self.head = temp
      self.head.nxt = self.head
      self.head.prev = self.head
      return

    if v < self.head.v:
      temp.nxt = self.head
      temp.prev = self.head.prev
      self.head.prev.nxt = temp
      self.head.prev = temp
      self.head = temp
      return

    last = self.head
    current = self.head.nxt
    while current != self.head and current.v < v:
      last = current
      current = current.nxt
    temp.nxt = current
    temp.prev = last
    current.prev = temp
    last.nxt = temp

  def to_s(self):
    if not self.head:
      return '[]'

    ret = '[ %d'%(self.head.v)
    current = self.head.nxt
    if current != self.head:
      ret +=', '
      while current != self.head:
        ret += str(current.v)
        if current.nxt != self.head:
          ret += ', '
        current = current.nxt
    ret += ' ]'
    return ret

  def getIndex(self, i):
    if not self.head:
      return None
    if i == 0:
      return self.head

    goingUp = i > 0
    startAt = 1 if goingUp else -1
    startNode = self.head.nxt if goingUp else self.head
    moveOp = 1 if goingUp else -1
    while startAt != i:
      startAt += moveOp
      startNode = startNode.nxt if goingUp else startNode.prev
    return startNode

  def indexOf(self, v):
    if not self.head:
      return None
    if v == self.head:
      return 0

    index = 0
    current = self.head.nxt
    while current != self.head:
      index += 1
      if current.v == v:
        return index
      current = current.nxt
    return None

  def swapIndexes(self, first, second):
    if (first == second):
      return

    firstNode = self.getIndex(first)
    secondNode = self.getIndex(second)

    if (firstNode == secondNode):
      return

    self.swapNodes(firstNode, secondNode)

  def swapNodes(self, firstNode, secondNode):
    # temp values if not adjacent
    oldsnp = secondNode.prev
    oldfnn = firstNode.nxt
    adjacent = oldsnp == firstNode

    # start unhooking in same direction as temp values
    # (sn.p & fn.n, inversion is sn.n & fn.p)
    secondNode.prev = firstNode.prev
    secondNode.prev.nxt = secondNode
    firstNode.nxt = secondNode.nxt
    firstNode.nxt.prev = firstNode

    # adjacent means first node behind(prev) second node (..->fn->sn->..)
    # if adjancent, sn and fn should link directly (..->sn->fn->..)
    # otherwise use the temp values
    if (adjacent):
      secondNode.nxt = firstNode
      firstNode.prev = secondNode
    else:
      firstNode.prev = oldsnp
      firstNode.prev.nxt = firstNode
      secondNode.nxt = oldfnn
      secondNode.nxt.prev = secondNode

    if self.head == firstNode:
      self.head = secondNode
    elif self.head == secondNode:
      self.head = firstNode

  def insertAt(self, i, v):
    nodeAt = self.getIndex(i)
    if not nodeAt:
      self.addSorted(v)
      return

    temp = Node(v, nodeAt, nodeAt.prev)
    nodeAt.prev.nxt = temp
    nodeAt.prev = temp

    if i == 0:
      self.head = temp

  def clear(self):
    self.head = None

  def insertSort(self):
    if not self.head or self.head.nxt == self.head:
      return

    outerLast = self.head
    outerNext = outerLast.nxt
    while outerNext != self.head:
      # should we step back?
      if outerLast.v > outerNext.v:
        outerMirror = outerNext
        outerBackstep = outerLast
        while outerBackstep != self.head and outerMirror.v < outerBackstep.prev.v:
          outerBackstep = outerBackstep.prev


        # moving (om)30 past (ob)90 = 30->90
        # disconnect self
        outerMirror.prev.nxt = outerMirror.nxt
        outerMirror.nxt.prev = outerMirror.prev

        # update self
        outerMirror.nxt = outerBackstep
        outerMirror.prev = outerBackstep.prev

        # update around self
        outerMirror.prev.nxt = outerMirror
        outerMirror.nxt.prev = outerMirror

        if outerBackstep == self.head:
          self.head = outerMirror
      # iterate forward
      outerLast = outerNext
      outerNext = outerNext.nxt

  def tail(self):
    return self.head.prev if self.head else None

  def lastIndex(self):
    return self.indexOf(self.tail().v) if self.tail() else None

  def shuffle(self):
    if not self.head:
      return
    li = self.lastIndex()
    for i in xrange(0, li):
      outer = li - i
      swap = random.randint(0, outer)
      self.swapIndexes(outer, swap)

  def bubbleSort(self):
    if not self.head or self.head.nxt == self.head:
      return

    outer = self.tail()
    while outer != self.head:
      current = self.head
      while current != outer:
        if (current.nxt.v < current.v and current.nxt != self.head):
          self.swapNodes(current, current.nxt)
          if current.prev == outer:
            outer = current
        else:
          current = current.nxt
      outer = outer.prev


  def find(self, v):
    if not self.head:
      return None
    if self.head.v == v:
      return self.head

    current = self.head.nxt
    while current != self.head:
      if current.v == v:
        return current
      current = current.nxt
    return None

  def to_a(self):
    if not self.head:
      return []

    ret = [ self.head.v ]
    current = self.head.nxt
    while current != self.head:
      ret.append(current.v)
      current = current.nxt
    return ret

  def delete(self, v):
    node = self.find(v)
    if not node:
      return

    self.deleteNode(node)

  def deleteAt(self, i):
    node = self.getIndex(i)
    if not node:
      return

    self.deleteNode(node)

  def deleteNode(self, n):
    n.prev.nxt = n.nxt
    n.nxt.prev = n.prev
    if n == self.head:
      self.head = n.nxt

File: javascript/double_linked_list/double_linked_list.js
const Node = function (value, next, prev) {
  this.value = value;
  this.next = next;
  this.prev = prev;
};

const DoubleLinkedList = function () {
  this.head = null;
};

DoubleLinkedList.prototype.addSorted = function (v) {
  // no nodes
  if (!this.head) {
    this.head = new Node(v);
    this.head.next = this.head;
    this.head.prev = this.head;
    return;
  }
  let temp = new Node(v);

  // v is less than head
  if (this.head.value > v) {
    temp.next = this.head;
    temp.prev = this.head.prev;
    this.head.prev.next = temp;
    this.head.prev = temp;
    this.head = temp;
    return;
  }

  // find nodes that will surround
  let lastSmaller = this.head;
  let firstBigger = this.head.next;
  while(firstBigger != this.head && firstBigger.value < v) {
    lastSmaller = firstBigger;
    firstBigger = firstBigger.next;
  }
  temp.next = firstBigger;
  temp.prev = lastSmaller;
  lastSmaller.next = temp;
  firstBigger.prev = temp;
};

DoubleLinkedList.prototype.to_s = function () {
  if (!this.head) {
    return '[ ]';
  }

  let ret = '[ ' + this.head.value;
  let current = this.head.next;
  if (current != this.head) {
    ret += ', ';
  }
  while (current && current != this.head) {
    ret += current.value;
    if (current.next != this.head) {
      ret += ', ';
    }
    current = current.next;
  }
  ret += ' ]';
  return ret;
};

DoubleLinkedList.prototype.getIndex = function (i) {
  if (i == 0) {
    return this.head;
  }

  const moveRight = i > 0;
  const moveDelta = moveRight ? 1 : -1;
  let current = moveRight ? this.head : this.head.prev;
  let counter = moveRight ? 0 : -1;
  while (counter != i) {
    current = moveRight ? current.next : current.prev;
    counter += moveDelta;
  }
  return current;
};

DoubleLinkedList.prototype.indexOf = function (v) {
  if (!this.head) {
    return -1;
  }

  if (this.head.value == v) {
    return 0;
  }

  let counter = 1;
  let current = this.head.next;
  while (current != this.head) {
    if (current.value == v) {
      return counter;
    }
    current = current.next;
    counter++;
  }
  return -1;
};

DoubleLinkedList.prototype.swapIndexes = function (first, second) {
  if (first == second) {
    return;
  }

  const firstNode = this.getIndex(first);
  const secondNode = this.getIndex(second);

  this.swapNodes(firstNode, secondNode);
};

DoubleLinkedList.prototype.swapNodes = function (first, second) {

  // temp stores
  const snoldprev = second.prev;
  const fnoldnext = first.next;
  const adjacent = snoldprev == first;

  // first swap
  second.prev = first.prev;
  second.prev.next = second;
  first.next = second.next;
  first.next.prev = first;

  // checked adj swap
  if (adjacent) {
    second.next = first;
    first.prev = second;
  } else {
    second.next = fnoldnext;
    second.next.prev = second;
    first.prev = snoldprev;
    first.prev.next = first;
  }

  // checked new head
  if (this.head == first) {
    this.head = second;
  } else if (this.head == second) {
    this.head = first;
  }
};

DoubleLinkedList.prototype.insertAt = function (i, v) {
  // if no head, call generic add function
  if (!this.head) {
    this.addSorted(v);
    return;
  }

  const node = this.getIndex(i);
  let temp;

  if (i >= 0) {
    temp = new Node(v, node, node.prev);
    node.prev.next = temp;
    node.prev = temp;
    if (node == this.head && i == 0) {
      this.head = temp;
    }
  } else {
    // node = node.prev
    temp = new Node(v, node.next, node);
    node.next.prev = temp;
    node.next = temp;
  }
};

DoubleLinkedList.prototype.clear = function () {
  this.head = null;
};

DoubleLinkedList.prototype.insertSort = function () {
  // return if size is 0 or 1
  if (!this.head || !this.head.next) {
    return;
  }

  let outerCurrent = this.head.next;
  while (outerCurrent != this.head) {
    let backStep = outerCurrent.prev;
    if (backStep.value > outerCurrent.value) {
      // when we move outerCurrent, we lose the correct link to the next item
      // wherever outerCurrent moves, the next outerCurrent is always here
      const newOuter = outerCurrent.next;

      // step back until backStep is smaller than outerCurrent, or it's the head
      while(backStep != this.head && backStep.value > outerCurrent.value) {
        backStep = backStep.prev;
      }
      // release self
      outerCurrent.prev.next = outerCurrent.next;
      outerCurrent.next.prev = outerCurrent.prev;

      // if backStep == head and smaller, add before, else add after
      if (backStep == this.head && outerCurrent.value < backStep.value) {
        // update self
        outerCurrent.next = backStep;
        outerCurrent.prev = backStep.prev;

        // update surround
        backStep.prev.next = outerCurrent;
        backStep.prev = outerCurrent;

        // update head
        this.head = outerCurrent;
      } else {
        // update self
        outerCurrent.next = backStep.next;
        outerCurrent.prev = backStep;

        // update surround
        backStep.next.prev = outerCurrent;
        backStep.next = outerCurrent;
      }

      // move outerCurrent forward
      outerCurrent = newOuter;
    } else {
      outerCurrent = outerCurrent.next;
    }
  }
};

DoubleLinkedList.prototype.shuffle = function () {
  const lastIndex = this.indexOf(this.head.prev.value);
  for (let i = lastIndex; i > 0; i--) {
    const swapAt = Math.floor(Math.random() * i);
    this.swapIndexes(i, swapAt);
  }
};

DoubleLinkedList.prototype.bubbleSort = function () {
  if (!this.head || !this.head.next) {
    return;
  }

  let outer = this.head.prev;
  while (outer != this.head) {
    let current = this.head;
    while (current != outer) {
      if (current.next.value < current.value && current.next != this.head) {
        this.swapNodes(current, current.next);
        if (current.prev == outer) {
          outer = current;
        }
      } else {
        current = current.next;
      }
    }
    outer = outer.prev;
  }
};

DoubleLinkedList.prototype.find = function (v) {
  if (!this.head) {
    return null;
  }

  if (this.head.value == v) {
    return this.head;
  }

  let current = this.head.next;
  while (current != this.head) {
    if (current.value == v) {
      return current;
    }
    current = current.next;
  }
  return null;
};

DoubleLinkedList.prototype.deleteAt = function (i) {
  let toDelete = this.getIndex(i);

  if (!toDelete) {
    return false;
  }

  this.doDeletion(toDelete);
  return true;
};


DoubleLinkedList.prototype.delete = function (v) {
  let toDelete = this.find(v);

  if (!toDelete) {
    return false;

  }

  this.doDeletion(toDelete);
  return true;
};

DoubleLinkedList.prototype.doDeletion = function (node) {
  node.prev.next = node.next;
  node.next.prev = node.prev;

  if (node == this.head) {
    this.head = node.next;
  }
};

DoubleLinkedList.prototype.to_a = function () {
  if (!this.head) {
    return [];
  }

  let ret = [ this.head.value ];
  let current = this.head.next;
  while (current != this.head) {
    ret.push(current.value);
    current = current.next;
  }
  return ret;
};

DoubleLinkedList.prototype.tail = function () {
  if (!this.head) {
    return null;
  }
  return this.head.prev;
};

module.exports = DoubleLinkedList;